File: execution-plane/isthmus-runtime/image/lab/livebootstrap/policy.py
"""Explicit lab-only single public binary bind; production provider is unchanged."""
import base64
import hashlib
import re

from lab.host import LABEL
from lab.toolchain import BASE_ID, ENVS

TMPFS = {"/tmp": "rw,noexec,nosuid,nodev,size=64m,mode=1777",
         "/run": "rw,noexec,nosuid,nodev,size=16m,mode=1777",
         "/home/claude": "rw,noexec,nosuid,nodev,size=128m,mode=0700,uid=1000,gid=1000"}


def public_config(value):
    if (not isinstance(value, dict) or set(value) != {"trust_sha256", "ticket_public_key"}
            or not isinstance(value["trust_sha256"], str)
            or not re.fullmatch(r"[a-f0-9]{64}", value["trust_sha256"])
            or not isinstance(value["ticket_public_key"], str)):
        raise ValueError("live_public_config_rejected")
    try:
        key = base64.b64decode(value["ticket_public_key"] + "=", validate=True)
    except Exception:
        raise ValueError("live_public_config_rejected") from None
    if len(key) != 32 or base64.b64encode(key).decode().rstrip("=") != value["ticket_public_key"]:
        raise ValueError("live_public_config_rejected")
    return value


def envs(label, public):
    if label not in ("a", "b"):
        raise ValueError("live_slot_rejected")
    public_config(public)
    account = "account-live-" + label
    values = {
        "EXECUTION_ACCOUNT_HASH": hashlib.sha256(account.encode()).hexdigest()[:32],
        "EXECUTION_SLOT_ID": "slot-live-" + label, "EXECUTION_NODE_ID": "node-live",
        "EXECUTION_EPOCH": "1", "EXECUTION_RUNTIME_GENERATION": "1",
        "EXECUTION_IMAGE_DIGEST": BASE_ID, "EXECUTION_LISTEN_ADDRESS": "127.0.0.1:8093",
        "EXECUTION_HEALTHCHECK_ADDRESS": "127.0.0.1:8093",
        "EXECUTION_IDENTITY_DIRECTORY": "/run/execution/identity",
        "EXECUTION_RUNTIME_TRUST_FILE": "/run/execution/runtime-ca.pem",
        "EXECUTION_BOOTSTRAP_CA_SHA256": public["trust_sha256"],
        "EXECUTION_TICKET_PUBLIC_KEY": public["ticket_public_key"],
        "EXECUTION_UPSTREAM_BASE_URL": "https://api.anthropic.com",
        "EXECUTION_EGRESS_PROXY_URL": "http://host-agent.execution.internal:8094",
        "EXECUTION_ALLOW_FAKE_ACTIVATION": "true",
    }
    return ENVS + [key + "=" + value for key, value in values.items()]


def args(lab, label, public):
    name = lab.name + "-live-" + label
    values = ["create", "--name", name, "--hostname", name, "--label", LABEL + "=" + lab.name,
              "--network", "none", "--read-only", "--user", "1000:1000", "--cap-drop", "ALL",
              "--security-opt", "no-new-privileges", "--memory", "1g", "--memory-swap", "1g",
              "--cpus", "1", "--pids-limit", "128", "--ulimit", "core=0:0", "--ipc", "private",
              "--cgroupns", "private", "--restart", "no", "--log-driver", "local",
              "--log-opt", "max-size=5m", "--log-opt", "max-file=2", "--workdir", "/",
              "--entrypoint", "/usr/bin/timeout", "--volume", str(lab.root / "live-bin/worker") + ":/worker:ro"]
    for target, value in TMPFS.items():
        values += ["--tmpfs", target + ":" + value]
    for value in envs(label, public):
        values += ["--env", value]
    return values + [BASE_ID, "--kill-after=5", "90", "/worker"]


def validate(value, lab, label, public):
    h, c = value["HostConfig"], value["Config"]
    name = lab.name + "-live-" + label
    if (not re.fullmatch(r"[a-f0-9]{64}", value["Id"]) or value["Name"] != "/" + name
            or value["Image"] != BASE_ID or c["Image"] != BASE_ID
            or c["Hostname"] != name or c["User"] != "1000:1000"
            or c.get("Labels", {}).get(LABEL) != lab.name or sorted(c["Env"]) != sorted(envs(label, public))
            or c["Entrypoint"] != ["/usr/bin/timeout"] or c["Cmd"] != ["--kill-after=5", "90", "/worker"]
            or c["WorkingDir"] != "/"):
        raise ValueError("live_instance_identity_mismatch")
    if (h["Privileged"] or not h["ReadonlyRootfs"] or h["NetworkMode"] != "none"
            or h["Memory"] != 1024**3 or h["MemorySwap"] != 1024**3 or h["NanoCpus"] != 10**9
            or h["PidsLimit"] != 128 or h.get("CapAdd") or h["CapDrop"] != ["ALL"]
            or h["SecurityOpt"] != ["no-new-privileges"]
            or h.get("PortBindings") or h.get("Devices") or h.get("DeviceRequests")
            or h.get("PidMode") not in ("", "private") or h["IpcMode"] != "private"
            or h["CgroupnsMode"] != "private" or h["RestartPolicy"]["Name"] != "no"
            or h["Tmpfs"] != TMPFS or h["Ulimits"] != [{"Name": "core", "Hard": 0, "Soft": 0}]):
        raise ValueError("live_instance_isolation_mismatch")
    if h.get("Binds") != [str(lab.root / "live-bin/worker") + ":/worker:ro"] or h.get("Mounts"):
        raise ValueError("live_bind_request_mismatch")
    mounts = value["Mounts"]
    binds = [m for m in mounts if m["Type"] == "bind"]
    if (len(binds) != 1 or binds[0]["Source"] != str(lab.root / "live-bin/worker")
            or binds[0]["Destination"] != "/worker" or binds[0]["RW"]
            or any(m["Type"] not in ("bind", "tmpfs") for m in mounts)
            or any(m["Type"] == "tmpfs" and m["Destination"] not in TMPFS for m in mounts)):
        raise ValueError("live_bind_actual_mismatch")

File: execution-plane/test/dockerbootstrap/policy.go
package main

import (
	"context"
	"reflect"
	"strings"

	"github.com/Wei-Shaw/sub2api/execution-plane/internal/provider"
	"github.com/Wei-Shaw/sub2api/execution-plane/internal/provider/docker"
	"github.com/Wei-Shaw/sub2api/execution-plane/internal/runtimeidentity"
)

func binding(index int) runtimeidentity.Binding {
	suffix := []string{"a", "b"}[index]
	return runtimeidentity.Binding{AccountHash: provider.RuntimeAccountID("account-live-" + suffix), SlotID: "slot-live-" + suffix, NodeID: nodeID, Epoch: 1, Generation: 1}
}

var tmpfs = map[string]string{
	"/tmp":         "rw,noexec,nosuid,nodev,size=64m,mode=1777",
	"/run":         "rw,noexec,nosuid,nodev,size=16m,mode=1777",
	"/home/claude": "rw,noexec,nosuid,nodev,size=128m,mode=0700,uid=1000,gid=1000",
}

func expectedEnvironment(index int, public publicConfig) map[string]string {
	b := binding(index)
	return map[string]string{
		"HOME": "/home/claude", "USER": "claude", "LOGNAME": "claude", "PATH": "/usr/bin:/bin", "LANG": "C.UTF-8", "LC_ALL": "C.UTF-8",
		"CLAUDE_CONFIG_DIR": "/home/claude/.claude", "DISABLE_AUTOUPDATER": "1", "CLAUDE_CODE_DISABLE_NONESSENTIAL_TRAFFIC": "1", "BUN_INSTALL_CACHE_DIR": "/home/claude/.bun-cache",
		"EXECUTION_LISTEN_ADDRESS": "127.0.0.1:8093", "EXECUTION_HEALTHCHECK_ADDRESS": "127.0.0.1:8093", "EXECUTION_ACCOUNT_HASH": b.AccountHash,
		"EXECUTION_SLOT_ID": b.SlotID, "EXECUTION_NODE_ID": nodeID, "EXECUTION_EPOCH": "1", "EXECUTION_RUNTIME_GENERATION": "1", "EXECUTION_IMAGE_DIGEST": baseImage,
		"EXECUTION_IDENTITY_DIRECTORY": "/run/execution/identity", "EXECUTION_RUNTIME_TRUST_FILE": "/run/execution/runtime-ca.pem",
		"EXECUTION_BOOTSTRAP_CA_SHA256": public.TrustSHA256, "EXECUTION_TICKET_PUBLIC_KEY": public.TicketPublicKey,
		"EXECUTION_UPSTREAM_BASE_URL": "https://api.anthropic.com", "EXECUTION_EGRESS_PROXY_URL": "http://host-agent.execution.internal:8094", "EXECUTION_ALLOW_FAKE_ACTIVATION": "true",
	}
}

// Python Lab additionally checks fields absent from the production Engine
// projection (MemorySwap, core ulimit, entrypoint and command). Never mutate an
// inspect result to make the production provider accept this lab-only profile.
func validateContainer(c docker.Container, config initialInput, public publicConfig, cid string, index int) error {
	h := c.HostConfig
	name := config.Owner + "-live-" + []string{"a", "b"}[index]
	if c.ID != cid || !cidPattern.MatchString(cid) || c.Name != "/"+name || c.Config.Hostname != name || c.Image != baseImage || c.Config.Image != baseImage || c.Config.User != "1000:1000" || c.Config.Labels[ownerLabel] != config.Owner || !c.State.Running {
		return rejected
	}
	if h.Privileged || !h.ReadonlyRootfs || h.NetworkMode != "none" || h.Memory != 1<<30 || h.NanoCPUs != 1_000_000_000 || h.PidsLimit != 128 ||
		len(h.CapAdd) != 0 || !reflect.DeepEqual(h.CapDrop, []string{"ALL"}) || !reflect.DeepEqual(h.SecurityOpt, []string{"no-new-privileges"}) ||
		(h.PidMode != "" && h.PidMode != "private") || h.IpcMode != "private" || h.CgroupnsMode != "private" || h.UTSMode != "" || h.UsernsMode != "" ||
		h.RestartPolicy.Name != "no" || h.PublishAllPorts || len(h.PortBindings) != 0 || len(c.NetworkSettings.Ports) != 0 ||
		len(h.Mounts) != 0 || len(h.VolumesFrom) != 0 || len(h.Devices) != 0 || len(h.DeviceRequests) != 0 || len(h.DeviceCgroupRules) != 0 || len(h.Links) != 0 || len(h.ExtraHosts) != 0 ||
		!reflect.DeepEqual(h.Tmpfs, tmpfs) || !reflect.DeepEqual(h.Binds, []string{config.WorkerPath + ":/worker:ro"}) {
		return rejected
	}
	if len(c.NetworkSettings.Networks) != 1 {
		return rejected
	}
	endpoint, ok := c.NetworkSettings.Networks["none"]
	if !ok || endpoint.IPAddress != "" || endpoint.Gateway != "" || endpoint.GlobalIPv6Address != "" {
		return rejected
	}
	binds := 0
	seen := map[string]bool{}
	for _, mount := range c.Mounts {
		if seen[mount.Destination] {
			return rejected
		}
		seen[mount.Destination] = true
		if mount.Type == "bind" {
			if mount.Source != config.WorkerPath || mount.Destination != "/worker" || mount.RW || mount.Propagation != "rprivate" {
				return rejected
			}
			binds++
		} else if mount.Type != "tmpfs" || tmpfs[mount.Destination] == "" {
			return rejected
		}
	}
	if binds != 1 {
		return rejected
	}
	actual := map[string]string{}
	for _, item := range c.Config.Env {
		key, value, ok := strings.Cut(item, "=")
		if !ok {
			return rejected
		}
		if _, exists := actual[key]; exists {
			return rejected
		}
		actual[key] = value
	}
	if !reflect.DeepEqual(actual, expectedEnvironment(index, public)) {
		return rejected
	}
	return nil
}

type inspector interface {
	InspectContainer(context.Context, string) (docker.Container, error)
}

func inspect(ctx context.Context, engine inspector, c initialInput, public publicConfig, ids []string, index int) error {
	if ctx.Err() != nil || len(ids) != 2 || index < 0 || index > 1 {
		return rejected
	}
	value, err := engine.InspectContainer(ctx, ids[index])
	if err != nil || ctx.Err() != nil {
		return rejected
	}
	return validateContainer(value, c, public, ids[index], index)
}
